artdaq-core-mu2e/Data/CalorimeterDataDecoder.hh

```diff
@@ -3,53 +3,90 @@
 
 #include "artdaq-core-mu2e/Data/DTCDataDecoder.hh"
 
+#include "TRACE/tracemf.h"
+
 #include <messagefacility/MessageLogger/MessageLogger.h> // Putting this here so that Offline/DAQ/src/FragmentAna_module.cc can use it
 
 namespace mu2e {
-class CalorimeterDataDecoder : public DTCDataDecoder
-{
+  class CalorimeterDataDecoder : public DTCDataDecoder
+  {
   public:
 
-	CalorimeterDataDecoder() : DTCDataDecoder() {}
+    CalorimeterDataDecoder() : DTCDataDecoder() {}
+
+    CalorimeterDataDecoder(std::vector<uint8_t> data);
+
+    CalorimeterDataDecoder(DTCLib::DTC_SubEvent const& f);
+
+    //Class to swap pairs of 16-bit words and extract 12-bit words without memory buffers -- only applies to DEBUG data
+    class Data12bitReader {
+      private:
+        const uint16_t* dataPtr;
+
+      public:
+        Data12bitReader(const uint16_t* dataPtr) : dataPtr(dataPtr) {}
 
-	CalorimeterDataDecoder(std::vector<uint8_t> data);
+        uint16_t operator[](size_t index) const {
+          uint16_t temp;
+          size_t wordIndex1 = (index * 3) / 4;
+          size_t wordIndex2 = wordIndex1+1;
+          uint16_t word1 = dataPtr[wordIndex1%2==0?wordIndex1+1:wordIndex1-1];
+          uint16_t word2 = dataPtr[wordIndex2%2==0?wordIndex2+1:wordIndex2-1];
 
-    explicit CalorimeterDataDecoder(DTCLib::DTC_SubEvent const& f);
+          switch (index % 4) {
+            case 0:
+              temp = (word1 >> 4) & 0x0FFF;
+              break;
+            case 1:
+              temp = ((word1 & 0x000F) << 8) | ((word2 & 0xFF00) >> 8);
+              break;
+            case 2:
+              temp = ((word1 & 0x00FF) << 4) | ((word2 & 0xF000) >> 12);
+              break;
+            case 3:
+              temp = word2 & 0x0FFF;
+              break;
+          }
+          return temp;
+        }
+    };
 
-    // CalorimeterHitDataPacket: Each hit is readout as a variable length sequence of data packets
     struct CalorimeterHitDataPacket
     {
       uint16_t DetectorType : 3; // subdetector type e.g. CALO=0, CAPHRI = 1, TRAD = 2, LASER = 3
       uint16_t BoardID : 8; //unique board ID from 0 - 255
-      uint16_t ChannelNumber : 6; // channel ID from 0-19
+      uint16_t ChannelNumber : 5; // channel ID from 0-19
```

Comment:
Why has this value changed?

Reply:
I changed it to match [DocDB@4914](https://mu2e-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=4914):
![image](https://github.com/user-attachments/assets/169015c0-b6fa-44cf-b605-646a0e945c91)